src/vos/vos_obj.c

```diff
@@ -1960,70 +2199,160 @@ vos_obj_iter_fini(struct vos_iterator *iter)
 
-	case VOS_ITER_SINGLE:
-		return singv_iter_next(oiter);
+static int
+vos_obj_akey_iter_next(struct vos_iterator *iter, daos_anchor_t *anchor)
+{
+	struct vos_obj_iter *oiter = vos_iter2oiter(iter);
 
-	case VOS_ITER_RECX:
-		return recx_iter_next(oiter);
+	D_ASSERTF(iter->it_type == VOS_ITER_AKEY, "type is %d\n", iter->it_type);
+	if (is_fake_iter(oiter)) {
+		/** Indicate we are done iterating */
+		oiter->it_fake_akey = 0;
+		return -DER_NONEXIST;
 	}
+
+	return key_iter_next(oiter, anchor);
 }
 
 static int
-vos_obj_iter_fetch(struct vos_iterator *iter, vos_iter_entry_t *it_entry,
-		   daos_anchor_t *anchor)
+vos_obj_sv_iter_next(struct vos_iterator *iter, daos_anchor_t *anchor)
 {
 	struct vos_obj_iter *oiter = vos_iter2oiter(iter);
 
-	switch (iter->it_type) {
-	default:
-		D_ASSERT(0);
-		return -DER_INVAL;
+	D_ASSERTF(iter->it_type == VOS_ITER_SINGLE, "type is %d\n", iter->it_type);
+	return singv_iter_next(oiter);
+}
 
-	case VOS_ITER_DKEY:
-	case VOS_ITER_AKEY:
-		return key_iter_fetch(oiter, it_entry, anchor, false, 0);
+static int
+vos_obj_ev_iter_next(struct vos_iterator *iter, daos_anchor_t *anchor)
+{
+	struct vos_obj_iter *oiter = vos_iter2oiter(iter);
 
-	case VOS_ITER_SINGLE:
-		return singv_iter_fetch(oiter, it_entry, anchor);
+	D_ASSERTF(iter->it_type == VOS_ITER_RECX, "type is %d\n", iter->it_type);
+	return recx_iter_next(oiter);
+}
 
-	case VOS_ITER_RECX:
-		return recx_iter_fetch(oiter, it_entry, anchor);
+static int
+vos_obj_dkey_iter_fetch(struct vos_iterator *iter, vos_iter_entry_t *it_entry,
+			daos_anchor_t *anchor)
+{
+	struct vos_obj_iter *oiter = vos_iter2oiter(iter);
+
+	D_ASSERTF(iter->it_type == VOS_ITER_DKEY, "type is %d\n", iter->it_type);
+
+	return key_iter_fetch(oiter, it_entry, anchor, false, 0);
+}
+
+static int
+vos_obj_akey_iter_fetch(struct vos_iterator *iter, vos_iter_entry_t *it_entry,
+			daos_anchor_t *anchor)
+{
+	struct vos_obj_iter *oiter = vos_iter2oiter(iter);
+
+	D_ASSERTF(iter->it_type == VOS_ITER_AKEY, "type is %d\n", iter->it_type);
+	if (is_fake_iter(oiter)) {
+		D_ASSERTF(oiter->it_fake_akey == '0', "Must probe before fetch");
+		if (anchor != NULL)
+			vos_fake_anchor_create(anchor);
+		if (fake_iter_child_is_array(oiter))
+			it_entry->ie_child_type = VOS_ITER_RECX;
+		else
+			it_entry->ie_child_type = VOS_ITER_SINGLE;
+		it_entry->ie_vis_flags = VOS_VIS_FLAG_VISIBLE;
+		if (oiter->it_ilog_info.ii_create == 0) {
+			/* The key has no visible subtrees so mark it covered */
+			it_entry->ie_vis_flags = VOS_VIS_FLAG_COVERED;
+		}
+
+		it_entry->ie_epoch     = oiter->it_epr.epr_hi;
+		it_entry->ie_punch     = oiter->it_ilog_info.ii_next_punch;
+		it_entry->ie_obj_punch = oiter->it_obj->obj_ilog_info.ii_next_punch;
+		/** Use the dkey for this */
+		vos_ilog_last_update(&oiter->it_dkey_krec->kr_ilog, VOS_TS_TYPE_DKEY,
+				     &it_entry->ie_last_update, !!oiter->it_iter.it_for_sysdb);
+		d_iov_set(&it_entry->ie_key, &oiter->it_fake_akey, sizeof(oiter->it_fake_akey));
+
+		return 0;
```

Comment:
I don't quite follow this, when this '0' fake akey will be fetched by iterator and how will the iterator caller deal with it?

Reply:
It just treats it like any other iterator value.  The intention was to not really change the interface. For example, rebuild will still see this fake akey and will use it to fetch from remote and to update locally.

At some point, I may revisit this and remove it altogether but I didn't want to have to change a lot of code outside of vos to support this change.